src/lib/iptables.ts

```diff
@@ -1,66 +1,170 @@
-		.then(() => applyIptablesArgs(`-A ${rule}`));
+export type RuleAdaptor = (rule: Rule) => Promise<string>;
+export function getDefaultRuleAdaptor(): RuleAdaptor {
+	return iptables;
 }
 
-function clearAndInsertIptablesRule(rule: string): Bluebird<void> {
-	return clearIptablesRule(rule)
-		.catchReturn(null)
-		.then(() => applyIptablesArgs(`-I ${rule}`));
+const iptables: RuleAdaptor = async (rule: Rule) => {
+	const args: string[] = [];
+
+	if (rule.action) {
+		args.push(rule.action);
+	}
+	if (rule.chain) {
+		args.push(rule.chain);
+	}
+	if (rule.action === '-L') {
+		args.push('--line-numbers');
+		args.push('-n');
+	}
+	if (rule.table) {
+		args.push(`-t ${rule.table}`);
+	}
+	if (rule.proto) {
+		args.push(`-p ${rule.proto}`);
+	}
+	if (rule.matches) {
+		rule.matches.forEach((match) => args.push(match));
+	}
+	if (rule.comment) {
+		args.push('-m comment');
+		args.push(`--comment "${rule.comment}"`);
+	}
+	if (rule.target) {
+		args.push(`-j ${rule.target}`);
+	}
+
+	if (!rule.family) {
+		rule.family = 4;
+	}
+
+	const cmd = rule.family === 6 ? 'ip6tables' : 'iptables';
+	return new Promise<string>((resolve, reject) => {
+		const proc = child_process.spawn(cmd, args, { shell: true });
+
+		const stdout: string[] = [];
+		proc.stdout.on('data', (data: Buffer) => {
+			stdout.push(data.toString('utf8'));
+		});
+
+		const stderr: string[] = [];
+		proc.stderr.on('data', (data: Buffer) => {
+			stderr.push(data.toString('utf8'));
+		});
+
+		proc.on('error', (err) => reject(err));
+		proc.on('close', (code) => {
+			if (code && code !== 0) {
+				return reject(
+					new IPTablesRuleError(
+						`Error running iptables: ${stderr.join()} (${args.join(' ')})`,
+					),
+				);
+			}
+			return resolve(stdout.join());
+		});
+	});
+};
+
+export function forChain(chain: string, table?: string): RuleBuilder {
+	const builder = {
+		rules: new Array<Rule>(),
+		add: (r: Rule) => {
+			const newRules = _.castArray(r);
+			builder.rules.push(
+				...newRules.map((rule) => {
+					return {
+						...rule,
+						...{
+							chain,
+							table,
+						},
+					};
+				}),
+			);
+			return builder;
+		},
+		apply: async (adaptor: RuleAdaptor) => {
+			await applyRules(builder.rules, adaptor);
+		},
+	};
+	return builder;
 }
 
-export function rejectOnAllInterfacesExcept(
-	allowedInterfaces: string[],
-	port: number,
-): Bluebird<void> {
-	// We delete each rule and create it again to ensure ordering (all ACCEPTs before the REJECT/DROP).
-	// This is especially important after a supervisor update.
-	return Bluebird.each(allowedInterfaces, (iface) =>
-		clearAndInsertIptablesRule(
-			`INPUT -p tcp --dport ${port} -i ${iface} -j ACCEPT`,
-		),
-	)
-		.then(() =>
-			clearAndAppendIptablesRule(
-				`OUTPUT -p tcp --sport ${port} -m state --state ESTABLISHED -j ACCEPT`,
-			),
-		)
-		.then(() =>
-			clearAndAppendIptablesRule(`INPUT -p tcp --dport ${port} -j REJECT`),
+async function applyRule(rule: Rule, adaptor: RuleAdaptor) {
```

Comment:
Can you either split this function or add some comments please? Following what's going on here isn't too easy.